doc/designs/quic-design/connection-state-machine.md

```diff
@@ -0,0 +1,637 @@
+     this state, a packet containing a `CONNECTION_CLOSE` frame is
+     transmitted again in response to any packets received. This ensures
+     that a `CONNECTION_CLOSE` frame is received by the peer even if the
+     initially transmitted `CONNECTION_CLOSE` frame was lost. Note that
+     these `CONNECTION_CLOSE` frames are not governed by QUIC's normal loss
+     detection mechanisms; this is a bespoke mechanism unique to this
+     state, which exists solely to ensure delivery of the `CONNECTION_CLOSE`
+     frame.
+
+     The endpoint progresses to the Terminated state after a timeout
+     interval, which should not be less than three times the PTO interval.
+
+     It is also possible for the endpoint to transition to the Draining
+     state instead, if it receives a `CONNECTION_CLOSE` frame prior
+     to the timeout expiring. This indicates that the peer is also
+     closing.
+
+   - The Draining state, used for a peer initiated immediate close.
+
+     The local endpoint may not send any packets of any kind in this
+     state. It may optionally send one `CONNECTION_CLOSE` frame immediately
+     prior to entering this state.
+
+     The endpoint progresses to the Terminated state after a timeout
+     interval, which should not be less than three times the PTO interval.
+
+- The Terminated state is the terminal state of a connection.
+  Regardless of how a connection ends (local or peer-initiated immediate close,
+  idle timeout, stateless reset), a connection always ultimately ends up in this
+  state. There is no longer any requirement to send or receive any packet. No
+  timer events related to the connection will ever need fire again. This is a
+  totally quiescent state. The state associated with the connection may now be
+  safely freed.
+
+We express this state machine in more concrete form in the form of a table,
+which makes the available transitions clear:
+
+† Except where superceded by a more specific transition
+
+ε means “where no other transition is applicable”.
+
+Where an action is specified in the Transition/Action column but no new state,
+no state change occurs.
+
+<table>
+<tr><th>State</th><th>Action On Entry/Exit</th><th>Event</th><th>Transition/Action</th></tr>
+<tr>
+  <td rowspan="2"><tt>IDLE</tt></td>
+  <td rowspan="2"></td>
+  <td>—<tt>APP:CONNECT</tt>→</td>
+  <td><tt>ACTIVE.ESTABLISHING.PROACTIVE_VER_NEG</tt> (if used), else
+  <tt>ACTIVE.ESTABLISHING.PRE_INITIAL</tt></td>
+</tr>
+<tr>
+  <td>—<tt>APP:CLOSE</tt>→</td>
+  <td><tt>TERMINATED</tt></td>
+</tr>
+<tr>
+  <td rowspan="5"><tt>ACTIVE</tt></td>
+  <td rowspan="5"></td>
+  <td>—<tt>IDLE_TIMEOUT</tt>→</td>
+  <td><tt>TERMINATED</tt></td>
+</tr>
+<tr>
+  <td>—<tt>PROBE_TIMEOUT</tt>→ †</td>
+  <td><tt>SendProbeIfAnySentPktsUnacked()</tt></td>
+</tr>
+<tr>
+  <td>—<tt>APP:CLOSE</tt>→ †</td>
+  <td><tt>TERMINATING.CLOSING</tt></td>
+</tr>
+<tr>
+  <td>—<tt>RX:ANY[CONNECTION_CLOSE]</tt>→</td>
```

Comment:
This event when seen on the diagram is confusingly placed. On the diagram it looks like this is coming from TERMINATING and going to TERMINATED.

Reply:
Unfortunately not much I can do about this, since the diagram is automatically generated.

Reply:
Ok - fair enough.